File: 1_SOLID_Design_Patterns/1.03_Liskov_Substitution_Principle/Python_Snippet/liskov_substitution_principle.py
# Create a birds that can have the following functions
# 1. Walking
# 2. Flying

# Violates Liskov Substitution Principle
class Bird(object):
    def fly():
        pass
    def walk():
        pass

class Parrot(Bird):
    def fly():
        print("Parrot is Flying")
    def walk():
        print("Parrot is Walking")

class Penguin(Bird):
    def fly():
        print("Penguin is Flying! Oops penguins can't fly")
    def walk():
        print("Penguin is Walking")

# it's break the principle of LSP. Penguin can not fly.

# Complies with Liskov Substitution Principle
class BirdSpecies(object):
    pass

class FlyingBird(BirdSpecies):
    def fly():
        pass

class WalkingBird(BirdSpecies):
    def walk():
        pass

class Parrot(FlyingBird,WalkingBird):
    def fly():
        print("Parrot is Flying")
    def walk():
        print("Parrot is Walking")

class Penguin(WalkingBird):
    def walk():
        print("Penguin is Walking")

File: 1_SOLID_Design_Patterns/1.03_Liskov_Substitution_Principle/Java_Snippet/liskov_substitution_principle.java
// Create a birds that can have the following functions
// 1. Walking
// 2. Flying

//  Violates Liskov Substitution Principle
interface Bird {
    public void fly();

    public void walk();
}

class Parrot implements Bird {
    public void fly() {
        System.out.println("Parrot is Flying");
    }

    public void walk() {
        System.out.println("Parrot is Walking");
    }
}

class Penguin implements Bird {
    public void fly() {
        System.out.println("Penguin is Flying! Oops penguins can't fly");
    }

    public void walk() {
        System.out.println("Penguin is Walking");
    }
}

// it's break the principle of LSP. Penguin can not fly.

// Complies with Liskov Substitution Principle
interface BirdSpecies {
    // to do;
}

interface FlyingBird extends BirdSpecies {
    public void fly();
}

interface WalkingBird extends BirdSpecies {
    public void walk();
}

class ParrotBird implements FlyingBird, WalkingBird {
    public void fly() {
        System.out.println("Parrot is Flying");
    }

    public void walk() {
        System.out.println("Parrot is Walking");
    }
}

class PenguinBird implements WalkingBird {
    public void walk() {
        System.out.println("Penguin is Walking");
    }
}